### PR TITLE
To support PostgreSQL v15

### PR DESCRIPTION
**Pull request by zwyan0**
To support PostgreSQL v15
- Change the Function name.　RelationOpenSmgr→RelationGetSmgr
- add nbtsort-15.c
- add new test to check UNIQUE NULLS DISTINCT
- update the github workflows version

### CODE REVIEW

File: lib/pg_strutil.c [hunk 1]
```diff
@@ -615,14 +670,32 @@ ParseFunction(const char *value, bool argistype)
...
-		Value  *v = lfirst(l);
+	/*
+	 * v15 removed Value node struct
+	 */
+
+	#if PG_VERSION_NUM >= 150000
+	
+		foreach (l, names)
+		{
+			String *v = lfirst(l);
```
[the remark this answers is not in the document]
Comment (mikecaat) [reply lib/pg_strutil.c:681]: Is lfirst_node(String, l) better?

**Review by mikecaat — COMMENTED** [review]
Thanks!

File: lib/pg_btree.c [hunk 2]
```diff
@@ -140,6 +142,7 @@ SpoolerOpen(Spooler *self,
 	self->relinfo->ri_RelationDesc = rel;
 	self->relinfo->ri_TrigDesc = NULL;	/* TRIGGER is not supported */
 	self->relinfo->ri_TrigInstrument = NULL;
+
```
[the remark this answers is not in the document]
Comment (zwyan0) [reply lib/pg_btree.c:145]: I deleted this line.

File: lib/pg_btree.c [hunk 3]
```diff
@@ -228,6 +231,7 @@ IndexSpoolBegin(ResultRelInfo *relinfo, bool enforceUnique)
 	int				numIndices = relinfo->ri_NumIndices;
 	RelationPtr		indices = relinfo->ri_IndexRelationDescs;
 	BTSpool		  **spools;
+
```
[the remark this answers is not in the document]
Comment (zwyan0) [reply lib/pg_btree.c:234]: I deleted this line.

File: lib/pg_btree.c [hunk 4]
```diff
@@ -715,7 +724,18 @@ _bt_mergeload(Spooler *self, BTWriteState *wstate, BTSpool *btspool, BTReader *b
...
-#if PG_VERSION_NUM >= 90100
+
+	/* 
+	 * v15 change RelationOpenSmgr name to RelationGetSmgr
+	 */
+
+#if PG_VERSION_NUM >= 150000
+	if (!RELATION_IS_LOCAL(wstate->index)&& !(wstate->index->rd_rel->relpersistence == RELPERSISTENCE_UNLOGGED))
+	{
+		RelationGetSmgr(wstate->index);
```
[the remark this answers is not in the document]
Comment (zwyan0) [reply lib/pg_btree.c:735]: I changed the code.

File: lib/pg_strutil.c [hunk 5]
```diff
@@ -382,32 +399,70 @@ GetNextArgument(const char *ptr, char **arg, Oid *argtype, const char **endptr,
...
-		{
-			/*
-			 * It might actually fit in int32. Probably only INT_MIN can
-			 * occur, but we'll code the test generally just to be sure.
-			 */
-			int32		val32 = (int32) val64;
-
-			if (val64 == (int64) val32)
-				*argtype = INT4OID;
+		/*
```
[the remark this answers is not in the document]
Comment (zwyan0) [reply lib/pg_strutil.c:402]: I deleted this comment.

File: lib/pg_strutil.c [hunk 6]
```diff
@@ -615,14 +670,32 @@ ParseFunction(const char *value, bool argistype)
 											  ret.argtypes)),
 				 errhint("Could not choose a best candidate function.")));
 
-	foreach (l, names)
-	{
-		Value  *v = lfirst(l);
+	/*
```
[the remark this answers is not in the document]
Comment (zwyan0) [reply lib/pg_strutil.c:673]: I deleted this comment.

File: lib/pg_strutil.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (zwyan0) [reply lib/pg_strutil.c:681]: I changed the code .

File: lib/pg_strutil.c [hunk 7]
```diff
@@ -382,32 +399,70 @@ GetNextArgument(const char *ptr, char **arg, Oid *argtype, const char **endptr,
...
-			int32		val32 = (int32) val64;
-
-			if (val64 == (int64) val32)
-				*argtype = INT4OID;
+		/*
+		 * v15 deleted int8.h and scanint8(),
+		 * so, used strtoi64() to instead of it
+		 */
+
+		#if PG_VERSION_NUM >= 150000
```
[the remark this answers is not in the document]
Comment (zwyan0) [reply lib/pg_strutil.c:407]: I deleted the indent.

File: lib/nbtree/nbtsort-15.c [hunk 8]
```diff
@@ -0,0 +1,2021 @@
+/*-------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (zwyan0) [reply lib/nbtree/nbtsort-15.c:1]: I will do it after PG 15 is released.

File: lib/nbtree/nbtsort-common.c [hunk 9]
```diff
@@ -10,7 +10,34 @@
  */
 
 
-#if PG_VERSION_NUM >= 140000
+#if PG_VERSION_NUM >= 150000
```
[the remark this answers is not in the document]
Comment (zwyan0) [reply lib/nbtree/nbtsort-common.c:13]: I changed the code.

[the remark this answers is not in the document]
Comment (mikecaat) [reply lib/nbtree/nbtsort-common.c]: Is the following enough?
```
_bt_spoolinit(Relation heap, Relation index, bool isunique,
#if PG_VERSION_NUM >= 150000
                      bool nulls_not_distinct, 
#endif
                      bool isdead)
```

File: lib/pg_btree.c [hunk 10]
```diff
@@ -242,7 +244,12 @@ IndexSpoolBegin(ResultRelInfo *relinfo, bool enforceUnique)
 			elog(DEBUG1, "pg_bulkload: spool \"%s\"",
 				RelationGetRelationName(indices[i]));
 
-#if PG_VERSION_NUM >= 90300
+#if PG_VERSION_NUM >= 150000
+			spools[i] = _bt_spoolinit(heapRel,indices[i],
```
Comment (mikecaat) [inline lib/pg_btree.c:248]: Is the following enough?
```
spools[i] = _bt_spoolinit(heapRel,indices[i],
					enforceUnique ? indices[i]->rd_index->indisunique: false,
#if PG_VERSION_NUM >= 150000
					indices[i]->rd_index->indnullsnotdistinct,
#endif
                                        false);
```

File: lib/pg_btree.c [hunk 11]
```diff
@@ -715,10 +722,15 @@ _bt_mergeload(Spooler *self, BTWriteState *wstate, BTSpool *btspool, BTReader *b
 	 * fsync those pages here, they might still not be on disk when the crash
 	 * occurs.
 	 */
+
```
Comment (mikecaat) [inline lib/pg_btree.c:725]: Unindented newline?

File: lib/pg_strutil.c [hunk 12]
```diff
@@ -151,8 +156,13 @@ int
 ParseInt32(char *value, int minValue)
 {
 	int32	i;
-	
-	i = pg_atoi(value, sizeof(int32), 0);
+
+#if PG_VERSION_NUM  >= 150000
```
Comment (mikecaat) [inline lib/pg_strutil.c:160]: two spaces after "NUM" ?

File: lib/pg_strutil.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (mikecaat) [reply lib/pg_strutil.c:638]: The comments remains in the latest file?

File: lib/pg_strutil.c (hunk 12 above)
Comment (zwyan0) [reply lib/pg_strutil.c]: I deleted one spaces.

File: lib/pg_btree.c (hunk 11 above)
Comment (zwyan0) [reply lib/pg_btree.c]: I deleted this newline.

File: lib/pg_btree.c (hunk 10 above)
Comment (zwyan0) [reply lib/pg_btree.c]: I changed zhe "if" block.

File: lib/pg_strutil.c [hunk 13]
```diff
@@ -615,14 +635,22 @@ ParseFunction(const char *value, bool argistype)
...
+	/*
+	 * v15 removed Value node struct
+	 */
 
-		pfree(strVal(v));
-		pfree(v);
-	}
-	list_free(names);
+	
+		foreach (l, names)
```
[the remark this answers is not in the document]
Comment (zwyan0) [reply lib/pg_strutil.c]: last time.I added a "if" block. So, also added a indent in block.
I unindent the block.

File: lib/pg_strutil.c [hunk 14]
```diff
@@ -382,32 +394,40 @@ GetNextArgument(const char *ptr, char **arg, Oid *argtype, const char **endptr,
...
-			/*
-			 * It might actually fit in int32. Probably only INT_MIN can
-			 * occur, but we'll code the test generally just to be sure.
-			 */
-			int32		val32 = (int32) val64;
-
-			if (val64 == (int64) val32)
-				*argtype = INT4OID;
+#if PG_VERSION_NUM >= 150000
+			errno = 0;
```
[the remark this answers is not in the document]
Comment (zwyan0) [reply lib/pg_strutil.c]: last time.I added a "if" block. So, also added a indent in block.
I unindent the block.

File: lib/pg_strutil.c [hunk 15]
```diff
@@ -382,32 +394,40 @@ GetNextArgument(const char *ptr, char **arg, Oid *argtype, const char **endptr,
...
+			val64 = strtoi64(str, &tmpendptr, 10);
+
+			if (errno == 0 && *tmpendptr == '\0')
+#else
+			/* could be an oversize integer as well as a float ... */
```
[the remark this answers is not in the document]
Comment (zwyan0) [reply lib/pg_strutil.c]: No, for all versions. So I changed location of comment.

File: lib/pg_strutil.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (zwyan0) [reply lib/pg_strutil.c]: Sorry, I forgot to deleted the comment.

File: lib/nbtree/nbtsort-common.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (zwyan0) [reply lib/nbtree/nbtsort-common.c]: I changed zhe "if" block.

File: bin/expected/load_csv_2.out [hunk 16]
```diff
@@ -0,0 +1,693 @@
...
+ 5 | five
+   | six
+   | seven
+(7 rows)
+
+-- set unique nulls not distinct
+\! pg_bulkload -d contrib_regression data/csv9.ctl -i data/data9.csv -o "VERBOSE=YES" -l results/csv10.log -P results/csv10.prs -u results/csv10.dup
+NOTICE: BULK LOAD START
+ERROR: query failed: ERROR:  could not create unique index "unique_idx2"
+DETAIL:  Key (i)=(null) is duplicated.
```
Comment (zwyan0) [inline bin/expected/load_csv_2.out:684]: Need to check zhe code.

File: lib/pg_btree.c [hunk 17]
```diff
@@ -718,7 +723,11 @@ _bt_mergeload(Spooler *self, BTWriteState *wstate, BTSpool *btspool, BTReader *b
 #if PG_VERSION_NUM >= 90100
 	if (!RELATION_IS_LOCAL(wstate->index)&& !(wstate->index->rd_rel->relpersistence == RELPERSISTENCE_UNLOGGED))
 	{
+	#if PG_VERSION_NUM >= 150000
+		RelationGetSmgr(wstate->index);
+	#else
 		RelationOpenSmgr(wstate->index);
+	#endif
```
Comment (zwyan0) [inline lib/pg_btree.c:730]: Need to unindent

File: lib/pg_strutil.c [hunk 18]
```diff
@@ -151,8 +156,13 @@ int
 ParseInt32(char *value, int minValue)
 {
 	int32	i;
-	
-	i = pg_atoi(value, sizeof(int32), 0);
+
+#if PG_VERSION_NUM >= 150000
+		i = pg_strtoint32(value);
+#else
+		i = pg_atoi(value, sizeof(int32), 0);
```
Comment (zwyan0) [inline lib/pg_strutil.c:163]: Need to unindent

File: lib/pg_strutil.c [hunk 19]
```diff
@@ -348,7 +358,9 @@ GetNextArgument(const char *ptr, char **arg, Oid *argtype, const char **endptr,
 		const char *startptr;
 		char	   *str;
 		int64		val64;
-
+#if PG_VERSION_NUM >= 150000
+			char       *tmpendptr;
```
Comment (zwyan0) [inline lib/pg_strutil.c:362]: Need to unindent

File: lib/pg_strutil.c [hunk 20]
```diff
@@ -348,7 +358,9 @@ GetNextArgument(const char *ptr, char **arg, Oid *argtype, const char **endptr,
 		const char *startptr;
 		char	   *str;
 		int64		val64;
-
```
Comment (zwyan0) [inline lib/pg_strutil.c:351]: revert the newline

File: lib/pg_strutil.c [hunk 21]
```diff
@@ -382,13 +394,21 @@ GetNextArgument(const char *ptr, char **arg, Oid *argtype, const char **endptr,
 		str = palloc(len + 2);
 		snprintf(str, len + 2, "%c%s", minus ? '-' : '+', startptr);
 
-		/* could be an oversize integer as well as a float ... */
+/* could be an oversize integer as well as a float ... */
```
Comment (zwyan0) [inline lib/pg_strutil.c:397]: Need to add indent

File: lib/pg_strutil.c [hunk 22]
```diff
@@ -382,13 +394,21 @@ GetNextArgument(const char *ptr, char **arg, Oid *argtype, const char **endptr,
...
+#endif
 		{
 			/*
-			 * It might actually fit in int32. Probably only INT_MIN can
-			 * occur, but we'll code the test generally just to be sure.
-			 */
+			* It might actually fit in int32. Probably only INT_MIN can
+			* occur, but we'll code the test generally just to be sure.
+			*/
+
```
Comment (zwyan0) [inline lib/pg_strutil.c:411]: need to delete zhe newloine

File: lib/pg_strutil.c [hunk 23]
```diff
@@ -382,13 +394,21 @@ GetNextArgument(const char *ptr, char **arg, Oid *argtype, const char **endptr,
...
 		if (scanint8(str, true, &val64))
+#endif
 		{
 			/*
-			 * It might actually fit in int32. Probably only INT_MIN can
-			 * occur, but we'll code the test generally just to be sure.
-			 */
+			* It might actually fit in int32. Probably only INT_MIN can
+			* occur, but we'll code the test generally just to be sure.
+			*/
```
Comment (zwyan0) [inline lib/pg_strutil.c:410]: Need to add a space

File: lib/nbtree/nbtsort-common.c [hunk 24]
```diff
@@ -9,20 +9,26 @@
  *	  Copyright (c) 2021, NIPPON TELEGRAPH AND TELEPHONE CORPORATION
  */
 
-
-#if PG_VERSION_NUM >= 140000
 /*
  * create and initialize a spool structure
  */
+#if PG_VERSION_NUM >= 140000
```
[the remark this answers is not in the document]
Comment (zwyan0) [reply lib/nbtree/nbtsort-common.c]: I reverted the line.

File: lib/nbtree/nbtsort-common.c [hunk 25]
```diff
@@ -34,8 +40,11 @@ _bt_spoolinit(Relation heap, Relation index, bool isunique, bool isdead)
...
-													 btKbytes, NULL, false);
+#if PG_VERSION_NUM >= 150000
+													nulls_not_distinct,
+#endif
+													btKbytes, NULL, false);
 
 	return btspool;
 }
-#endif
+#endif
```
[the remark this answers is not in the document]
Comment (zwyan0) [reply lib/nbtree/nbtsort-common.c]: I added a newline

File: lib/pg_btree.c (hunk 17 above)
Comment (zwyan0) [reply lib/pg_btree.c]: I unindent this line.

File: lib/pg_strutil.c (hunk 19 above)
Comment (zwyan0) [reply lib/pg_strutil.c]: I unindent this line.

File: lib/pg_strutil.c (hunk 20 above)
Comment (zwyan0) [reply lib/pg_strutil.c:351]: I reverted this line

File: lib/pg_strutil.c (hunk 21 above)
Comment (zwyan0) [reply lib/pg_strutil.c]: I added a indent.

File: lib/pg_strutil.c (hunk 22 above)
Comment (zwyan0) [reply lib/pg_strutil.c]: I deleted this newline.

File: lib/pg_strutil.c (hunk 23 above)
Comment (zwyan0) [reply lib/pg_strutil.c:411]: I made a mistake in the correction. There should be a space in the bottom three lines.

File: .github/workflows/test.yml [hunk 26]
```diff
@@ -17,6 +17,7 @@ jobs:
     strategy:
       matrix:
         PGVERSION:                # TODO: build with master branch
+          - "15beta2"
```
Comment (mikecaat) [inline .github/workflows/test.yml:20]: need to fix and update minor version.

File: lib/nbtree/nbtsort-15.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (zwyan0) [reply lib/nbtree/nbtsort-15.c:1]: I compared it with the v15.1 file using diff command. There are no different.

File: .github/workflows/test.yml (hunk 26 above)
Comment (zwyan0) [reply .github/workflows/test.yml:20]: I updated it to the lastest version.

File: SPECS/pg_bulkload-pg15.spec [hunk 27]
```diff
@@ -0,0 +1,104 @@
...
+
+
+%description
+pg_bulkload provides high-speed data loading capability to PostgreSQL users.
+
+When we load huge amount of data to a database, it is common situation that data set to be loaded is valid and consistent. For example, dedicated tools are used to prepare such data, providing data validation in advance. In such cases, we'd like to bypass any overheads within database system to load data as quickly as possible. pg_bulkload is developed to help such situations. Therefore, it is not pg_bulkload's goal to provide detailed data validation. Rather, pg_bulkload asumes that loaded data set is validated by separate means. If you're not in such situation, you should use COPY command in PostgreSQL.
+
+
+%package llvmjit
+Requires: postgresql14-server, postgresql14-llvmjit
```
Comment (mikecaat) [inline SPECS/pg_bulkload-pg15.spec:33]: 15 is correct?

Comment (zwyan0) [reply SPECS/pg_bulkload-pg15.spec:33]: Sorry, 15 is correct. I will chang the package name.

**Review by mikecaat — APPROVED** [review]

**Pull request merged**